src/java.base/share/classes/java/nio/Buffer.java

```diff
@@ -823,16 +825,31 @@ public MemorySegment bufferSegment(Buffer buffer) {
                 }
 
                 @Override
-                public Runnable acquireSession(Buffer buffer, boolean async) {
-                    var session = buffer.session();
-                    if (session == null) {
+                public MemorySessionImpl acquireSession(Buffer buffer) {
+                    var scope = buffer.session();
+                    if (scope == null) {
                         return null;
                     }
-                    if (async && session.ownerThread() != null) {
-                        throw new IllegalStateException("Confined session not supported");
+                    scope.acquire0();
+                    return scope;
+                }
+
+                @Override
+                public void releaseSession(Buffer buffer, MemorySessionImpl scope) {
```

Comment:
I prefer methods that do not expose the scope implementation so access is limited to just to the acquire/release methods, but i am unsure of the performance implications. These methods might not reliably inline, and we may need to ensure that first (which is also separately a good thing). I think it requires that the shared secret fields are stable and that there is only one implementation of `JavaNioAccess`, which there is, but we can enforce via sealing. Something to consider as a further iteration.

Reply:
> I prefer methods that do not expose the scope implementation so access is limited to just to the acquire/release methods, but i am unsure of the performance implications. These methods might not reliably inline, and we may need to ensure that first (which is also separately a good thing). I think it requires that the shared secret fields are stable and that there is only one implementation of `JavaNioAccess`, which there is, but we can enforce via sealing. Something to consider as a further iteration.

I agree. A similar approach was done in a previous incarnation of the PR whereby we exposed a `ScopeAcqusition` that held an internal copy of the buffer and provided means to release. We feared that would cause performance regression. But your comment made me think of another way of doing acquire/release which would not expose the session and that would always be performant. I will explore this idea a bit more.